tests/poll.rs

```diff
@@ -201,6 +205,98 @@ fn registry_behind_arc() {
     handle2.join().unwrap();
 }
 
+#[test]
+fn registry_operations_are_thread_safe() {
+    init();
+
+    let mut poll = Poll::new().unwrap();
+    let registry = Arc::new(poll.registry().try_clone().unwrap());
+    let mut events = Events::with_capacity(128);
+
+    let mut listener1 = TcpListener::bind(any_local_address()).unwrap();
+    let addr = listener1.local_addr().unwrap();
+    let mut listener2 = TcpListener::bind(any_local_address()).unwrap();
+    let barrier = Arc::new(Barrier::new(4));
+
+    let registry1 = Arc::clone(&registry);
+    let registry2 = Arc::clone(&registry);
+    let registry3 = Arc::clone(&registry);
+    let barrier1 = Arc::clone(&barrier);
+    let barrier2 = Arc::clone(&barrier);
+    let barrier3 = Arc::clone(&barrier);
+
+    let handle1 = thread::spawn(move || {
+        registry1
+            .register(&mut listener1, ID1, Interest::READABLE)
+            .unwrap();
+        registry1.deregister(&mut listener1).unwrap();
+        registry1
+            .register(&mut listener1, ID1, Interest::READABLE)
+            .unwrap();
+        registry1
+            .reregister(
+                &mut listener1,
+                ID1,
+                Interest::READABLE.add(Interest::WRITABLE),
+            )
+            .unwrap();
+
+        barrier1.wait();
+    });
+
+    let handle2 = thread::spawn(move || {
+        registry2
+            .register(&mut listener2, ID2, Interest::READABLE)
+            .unwrap();
+        registry2.deregister(&mut listener2).unwrap();
+        registry2
+            .register(&mut listener2, ID2, Interest::READABLE)
+            .unwrap();
+        registry2
+            .reregister(
+                &mut listener2,
+                ID2,
+                Interest::READABLE.add(Interest::WRITABLE),
+            )
+            .unwrap();
+
+        barrier2.wait();
+    });
+
+    let handle3 = thread::spawn(move || {
+        let mut stream = TcpStream::connect(addr).unwrap();
+        registry3
+            .register(&mut stream, ID3, Interest::READABLE.add(Interest::WRITABLE))
+            .unwrap();
+        registry3.deregister(&mut stream).unwrap();
+        registry3
+            .register(&mut stream, ID3, Interest::READABLE)
+            .unwrap();
+        registry3
+            .reregister(&mut stream, ID3, Interest::READABLE | Interest::WRITABLE)
+            .unwrap();
+
+        barrier3.wait();
+    });
+
+    std::thread::sleep(std::time::Duration::from_millis(1000));
```

Comment:
Is this sleep needed? `expect_events` below already loops/polls three time.

Reply:
If the timeout is too short you can use `barrier.wait()` before the call to `expect_events` (and call it twice at the of the threads above), to allow the threads to do there work.

Reply:
Thanks for the idea. Will use an extra barrier.wait call since it express better what I want to achieve here.